---
title: "shinysetup"
output: html_document
date: "2023-09-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(shiny)
library(tidyverse)
library(shinythemes)
library(DT)
library(plotly)


library(tidyverse)
library(DESeq2)
library(org.Mm.eg.db)
library(KEGGREST)
library(biomaRt)
library(gghalves)
library(ggpubr)
select <- dplyr::select

source("global.R", local = knitr::knit_global())
```


```{r import}
remove_list = c("IKOHL_1")

metadata = read_csv("metadata.csv")%>%
  column_to_rownames("Sample_ID")%>%
  mutate(genotype = fct_relevel(genotype, c("WT","FF","KO","IKO")))%>%
  mutate(bile_acid = fct_relevel(bile_acid, c("low","high")))%>%
  filter(!rownames(.) %in% remove_list)%>%
  mutate(geno = case_when(genotype %in% c("FF","IKO") ~ "Intestine",
                                          genotype %in% c("WT","KO") ~ "Global"))


countdata = read_delim("featurecount3.txt", skip = 1)%>%
  column_to_rownames("Geneid")%>%
  select(-(1:5))%>%
  rownames_to_column("rowname")%>%
  mutate(rowname = gsub("\\..*","",rowname))%>%
  column_to_rownames("rowname")

colnames(countdata) <- gsub("\\./aligned/+|\\.bam", "", colnames(countdata))

countdata = select(countdata, -remove_list)

```

```{r annotation}
ensembl <- useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://nov2020.archive.ensembl.org")

ann_df = getBM(attributes = c("ensembl_gene_id", "external_gene_name","description", "entrezgene_id"),
          filters = "ensembl_gene_id",
          values = rownames(countdata),
          mart = ensembl)


ann_dict <- setNames(ann_df$external_gene_name, ann_df$ensembl_gene_id)
  
write_csv(ann_df,"ann.csv")

ann_dict <- ann_dict[!is.na(ann_dict)]


count_named = countdata%>%
  rownames_to_column("ensembl_gene_id")%>%
  left_join(ann_df, by = "ensembl_gene_id")%>%
  select(ensembl_gene_id, external_gene_name, everything())
```


```{r normalize}
dds <-  DESeqDataSetFromMatrix(countData = countdata,
                               colData = metadata,
                               design = ~b_genotype + diet + diet:b_genotype)
dds <- estimateSizeFactors(dds)
countdata_n <-  counts(dds, normalized = TRUE)

count_named = countdata_n%>%
  as.data.frame()%>%
  rownames_to_column("ensembl_gene_id")%>%
  inner_join(ann_df, by = "ensembl_gene_id")%>%
  select(ensembl_gene_id, external_gene_name, everything())


write_csv(count_named, "counttable_n.csv")
```

```{r}
metadata_2 <- metadata%>%
  rownames_to_column("key")

gene_gg <- count_named%>%
  filter(external_gene_name == "Vps13a")%>%
  select(-external_gene_name, -ensembl_gene_id, -description,-entrezgene_id)%>%
  gather()%>%
  left_join(metadata_2)%>%
  mutate(value = as.numeric(value))



comparisons = list(c("WTL","KOH"), c("WTL", "KOL"), c("KOL", "KOH"))

ggplot(gene_gg, aes(x = b_genotype, y = value, color = b_genotype))+
  geom_half_violin(side = "l", aes(fill = b_genotype))+
  geom_boxplot(position = position_nudge(x = 0.22), width=0.2)+
  geom_half_point(side = "r")+
  facet_grid(geno~diet)+
  scale_fill_manual(values=c("#4E79A7", "#59A14F", "#F28E2B"),
                    labels = c("WT Low BA", "M2KO Low BA", "M2KO High BA"))+
  scale_color_manual(values=c("#4E79A7", "#59A14F", "#F28E2B"))+
  guides(color = "none")+
  stat_compare_means(comparisons = comparisons, method = "wilcox.test", hide.ns = TRUE)



plot_boxes("Vps13a")
```

```{r heatmap}
write_csv(msig_df, "msig_df.csv")
unique(msig_df$gs_name)


write.csv(dds_vst2, "dds_matrix.csv")

test_m <- read.csv("dds_matrix.csv",row.names = 1)


all(dds_vst2 == test_m)
```


```{r test heatmap}
generate_path_gene_list("HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION")
generate_heatmap(process_dds_vst(generate_path_gene_list("HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION")))

count_named%>%
    filter(str_detect(external_gene_name, regex(paste0("\\b", "1700047N06Rik", "\\b"), ignore_case = TRUE)))

count_named%>%
  select(external_gene_name)%>%
  filter(grepl("^[0-9]+[A-Za-z].*$", external_gene_name))
```

